File: Swift/프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.swift
import Foundation

func solution(_ n:Int) -> [[Int]] {
    var matrix = Array(repeating: Array(repeating: 0, count: n), count: n)
    let directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    var directionIndex = 0
    
    var row = 0
    var col = 0
    for num in 1...(n*n) {
        matrix[row][col] = num
        
        var nextRow = row + directions[directionIndex].0
        var nextCol = col + directions[directionIndex].1
        
        if !(0..<n).contains(nextRow) || !(0..<n).contains(nextCol) || matrix[nextRow][nextCol] != 0 {
            directionIndex = (directionIndex + 1) % 4
            
            nextRow = row + directions[directionIndex].0
            nextCol = col + directions[directionIndex].1
        }
        
        row = nextRow
        col = nextCol
    }
    
    return matrix
}

File: Python3/프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    matrix = [[0]*n for _ in range(n)]
    
    # 오른쪽, 아래, 왼쪽, 위
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    direction_index = 0
    
    row, col = 0, 0
    for num in range(1, n**2 + 1):
        matrix[row][col] = num
        
        next_row = row + directions[direction_index][0]
        next_col = col + directions[direction_index][1]
        
        if not(0 <= next_row < n and 0 <= next_col < n and matrix[next_row][next_col] == 0):
            direction_index = (direction_index + 1) % 4
            next_row = row + directions[direction_index][0]
            next_col = col + directions[direction_index][1]
        
        row, col = next_row, next_col
        
    return matrix
        
    